whelk-core/src/main/groovy/whelk/search2/querytree/Condition.java

```diff
@@ -108,77 +103,135 @@ public RdfSubjectType rdfSubjectType() {
+        List<Condition> prefilledFields = new ArrayList<>();
+        List<Selector> currentPath = new ArrayList<>();
+        for (Selector s : path) {
+            currentPath.add(s);
+            if (s instanceof Property p && p.isRestrictedSubProperty() && !p.hasIndexKey()) {
+                for (Restrictions.OnProperty r : p.objectOnPropertyRestrictions()) {
+                    // Support only HasValue restriction for now
+                    if (r instanceof Restrictions.HasValue(Property property, Value v)) {
+                        var restrictedPath = new Path(Stream.concat(currentPath.stream(), property.expand(jsonLd).path().stream()).toList());
+                        prefilledFields.add(new Condition(restrictedPath, EQUALS, v));
+                    }
+                }
+            }
+        }
+        return prefilledFields;
     }
 
-    private Optional<Map<String, Object>> getEsNestedQuery(ESSettings esSettings) {
-        return path.getEsNestedStem(esSettings.mappings())
-                .map(nestedStem -> nestedWrap(nestedStem, getCoreEsQuery(esSettings)));
+    // When querying type, match any subclass by default (TODO: make this optional)
+    private Node expandType(JsonLd jsonLd) {
+        if (!(selector.isType() && value instanceof VocabTerm v)) {
+            return this;
+        }
+
+        String baseType = v.key();
+
+        Set<String> subtypes = jsonLd.getSubClasses(baseType);
+        if (subtypes.isEmpty()) {
+            return this;
+        }
+
+        List<Node> altFields = Stream.concat(Stream.of(baseType), subtypes.stream())
+                .sorted()
+                .map(t -> (Node) withValue(new VocabTerm(t, jsonLd.vocabIndex.get(t))))
+                .toList();
+
+        return new Or(altFields);
+    }
+
+    private Map<String, Object> _toSearchMapping(Function<Node, Map<String, String>> makeUpLink) {
+        Map<String, Object> m = new LinkedHashMap<>();
+
+        m.put("property", selector.definition());
+        m.put(operator.termKey, value instanceof Resource r ? r.description() : value.queryForm());
+        m.put("up", makeUpLink.apply(this));
+
+        m.put("_key", selector.queryKey());
+        m.put("_value", value.queryForm());
+
+        return m;
     }
 
     private Map<String, Object> _getCoreEsQuery(String f, Value v, ESSettings esSettings) {
         if (operator.isRange() && !v.isRangeOpCompatible()) {
             // FIXME
             return nonsenseFilter();
         }
-        boolean valueIsObject = path.last() instanceof Property p && p.isObjectProperty();
         return switch (v) {
             case DateTime dateTime -> esDateFilter(f, dateTime, esSettings);
             case FreeText ft -> ft.isWild()
                     ? existsFilter(f)
-                    : esFreeTextFilter(valueIsObject ? f + "." + SEARCH_KEY : f, ft, esSettings);
+                    : esFreeTextFilter(selector.valueIsObject() ? f + "." + SEARCH_KEY : f, ft, esSettings);
```

Comment:
Maybe call it just `isObjectProperty` or `isObjectPropertySelector` in `Selector` as well.
As it's more about what it "accepts" than what the value is.